torch/distributed/algorithms/quantization.py

```diff
@@ -0,0 +1,117 @@
+            raise RuntimeError(
+                "Invalid function argument. Expected tensor "
+                "to be a single tensor list  of type torch.Tensor."
+            )
+        if (qtype == QuantType.FP16):
+            return self.fp32_to_fp16_with_clamp(tensor)
+        else:
+            raise RuntimeError(
+                'Quantization type {} is not supported'.format(qtype)
+            )
+
+    def quntize_tensor_list(self, tensor_list, qtype):
+        if not isinstance(tensor_list, list) or not all(
+            isinstance(p, torch.Tensor) for p in tensor_list
+        ):
+            raise RuntimeError(
+                "Invalid function argument. Expected "
+                "to be of type List[torch.Tensor]."
+            )
+        if (qtype == QuantType.FP16):
+            input_tensor = [self.quntize_tensor(t, qtype) for t in tensor_list]
+            return input_tensor
+        else:
+            raise RuntimeError(
+                'Quantization type {} is not supported'.format(qtype)
+            )
+
+    def dequantize_tensor(self, tensor, qtype, quant_loss=None):
+        if not isinstance(tensor, torch.Tensor):
+            raise RuntimeError(
+                "Invalid function argument. Expected "
+                "to be a single tensor list  of type torch.Tensor."
+            )
+        if (qtype == QuantType.FP16):
+            if tensor.dtype != torch.float16:
+                raise RuntimeError(
+                    "tensor dtype is {} while expected to be FP16.".format(
+                        tensor.dtype
+                    )
+                )
+            elif tensor.dtype == torch.float16 and quant_loss is None:
+                return tensor.float()
+            else:
+                return tensor.float() / quant_loss
+        else:
+            raise RuntimeError(
+                'Quantization type {} is not supported'.format(qtype)
+            )
+
+
+    def dequantize_tensor_list(self, tensors, qtype, quant_loss=None):
+        if not isinstance(tensors, list) or not all(
+            isinstance(p, torch.Tensor) for p in tensors
+        ):
+            raise RuntimeError(
+                "Invalid function argument. Expected "
+                "to be of type List[torch.Tensor]."
+            )
+        elif (qtype == QuantType.FP16):
+            tensor = [self.dequantize_tensor(t, qtype) for t in tensors]
+            return tensor
+        else:
+            raise RuntimeError(
+                'Quantization type {} is not supported'.format(qtype)
+            )
+
+
+def auto_quantize(func, qtype, quant_loss=None):
+    def wrapper(tensors, *args, **kwargs):
+        group = kwargs.get('group', None)
+        async_op = kwargs.get('async_op', False)
+        quant = quantization_wrapper()
+        if (func == dist.all_gather):
+            input_tensors = quant.quntize_tensor(args[0], qtype)
+            out_tensors = quant.quntize_tensor_list(tensors, qtype)
```

Comment:
also, making the first argument of `wrapper` with `tensors` are limiting the first argument type still? For example `dist.all_to_all_single` we have the first argument as a tensor type instead of a tensor list, so we can't for sure the first argument is always a tensor list. 

why not we just make wrapper signature be like `def wrapper(*args, **kwargs)` and extract the arguments base on the `func` type, i.e. all_gather we check if args[0] is a list, if it is we quantize_tensor_list, if it's tensor we quantize tensor.